### PR TITLE
Add infrastructure to test WASM linking

### PR DESCRIPTION
**Pull request by pranavkm**
The intent is to have a project that links Blazor assemblies and provide trimming analysis. This causes the build to fail if there's an API that isn't trimmer-safe or has been annotated or suppressed. Based on the pattern that the runtime uses - https://github.com/dotnet/runtime/blob/master/src/libraries/illink-sharedframework.targets

* [x] Annotate Microsoft.AspNetCore APIs that are part of Blazor WASM
* [x] Add a project to verify no trimmer warnings are produced from Microsoft.AspNetCore.* assemblies
* [x] Add trimmer configure to workaround issues with trimming Microsoft.AspNetCore assemblies
* [ ] TBD: Update SDK to trim these assemblies like normal.
-----

Missing the Auth assemblies, but I was able to verify that the others work fine. Note that we have no E2E integration for this as yet since the list of assemblies to type-granular trim is in the SDK which hasn't been modified outside of my dev box.

### CODE REVIEW

File: src/Components/Components/src/ParameterView.cs [hunk 1]
```diff
@@ -232,7 +232,7 @@ public static ParameterView FromDictionary(IDictionary<string, object?> paramete
         /// match the corresponding entry in the <see cref="ParameterView"/>.
         /// </summary>
         /// <param name="target">An object that has a public writable property matching each parameter's name and type.</param>
-        public void SetParameterProperties(object target)
+        public void SetParameterProperties([DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.PublicProperties | DynamicallyAccessedMemberTypes.NonPublicProperties)] object target)
```
Comment (davidfowl) [inline src/Components/Components/src/ParameterView.cs:235]: This doesn't work

**Review by dougbu — COMMENTED** [review]
Anything in particular you need @dotnet/aspnet-build to comment on @pranavkm❔ sln and eng/targets changes look fine, modulo one comment&hellip;

File: src/Components/Components/src/Properties/ILLink.Descriptors.xml [hunk 2]
```diff
@@ -0,0 +1,11 @@
+<linker>
+  <assembly fullname="Microsoft.AspNetCore.Components">
+    <type fullname="Microsoft.AspNetCore.Components.ComponentBase" required="false" />
```
Comment (eerhardt) [inline src/Components/Components/src/Properties/ILLink.Descriptors.xml:3]: I'm not sure I follow this. Is this "type level" granularity trimming?

File: src/Components/Components/src/Reflection/MemberAssignment.cs [hunk 3]
```diff
@@ -11,18 +12,23 @@ namespace Microsoft.AspNetCore.Components.Reflection
...
         {
             var dictionary = new Dictionary<string, List<PropertyInfo>>();
 
             Type? currentType = type;
 
             while (currentType != null)
             {
-                var properties = currentType.GetProperties(bindingFlags)
-                    .Where(prop => prop.DeclaringType == currentType);
+                var properties = currentType.GetProperties(bindingFlags);
```
Comment (eerhardt) [inline src/Components/Components/src/Reflection/MemberAssignment.cs:24]: Can you just add `bindingFlags | BindingFlags.Declared`?

File: src/Components/Components/src/Rendering/RenderTreeBuilder.cs [hunk 4]
```diff
@@ -4,7 +4,9 @@
 using System;
 using System.Collections.Generic;
 using System.Diagnostics;
+using System.Diagnostics.CodeAnalysis;
```
Comment (eerhardt) [inline src/Components/Components/src/Rendering/RenderTreeBuilder.cs:7]: Necessary?

File: src/Components/Components/src/CascadingParameterState.cs [hunk 5]
```diff
@@ -24,6 +24,7 @@ public CascadingParameterState(string localValueName, ICascadingValueComponent v
             ValueSupplier = valueSupplier;
         }
 
+        [UnconditionalSuppressMessage("ReflectionAnalysis", "IL2072:RequiresUnreferencedCode", Justification = "Requires a gesture that ensures components are always preserved. https://github.com/mono/linker/issues/1806")]
```
Comment (eerhardt) [inline src/Components/Components/src/CascadingParameterState.cs:27]: I wonder if it would be better to baseline these in an .xml file, like we do in dotnet/runtime. That way we know to come back to them when the linker feature is done.

The benefit is that the application will still get a warning, because what this is doing is unsafe.

File: src/Components/Components/src/DefaultComponentActivator.cs [hunk 6]
```diff
@@ -10,10 +11,11 @@ internal class DefaultComponentActivator : IComponentActivator
         public static IComponentActivator Instance { get; } = new DefaultComponentActivator();
 
         /// <inheritdoc />
+        [UnconditionalSuppressMessage("ReflectionAnalysis", "IL2067", Justification = "Requires a gesture that ensures components are always preserved. https://github.com/mono/linker/issues/1806")]
         public IComponent CreateInstance(Type componentType)
```
Comment (eerhardt) [inline src/Components/Components/src/DefaultComponentActivator.cs:15]: You should flow `DynamicallyAccessedMembers.ParameterlessConstructor` annotation to this type. And if necessary, suppress it upstream.

File: src/Components/Forms/src/EditContextDataAnnotationsExtensions.cs [hunk 7]
```diff
@@ -94,6 +94,7 @@ private static void ValidateField(EditContext editContext, ValidationMessageStor
             }
         }
 
+        [UnconditionalSuppressMessage("ReflectionAnalysis", "IL2080", Justification = "This is not a trimmer safe pattern.")]
```
Comment (eerhardt) [inline src/Components/Forms/src/EditContextDataAnnotationsExtensions.cs:97]: This shouldn't be suppressed. But instead have `RequiresUnreferencedCode` attribute.

File: src/Components/Shared/src/ComponentParametersTypeCache.cs [hunk 8]
```diff
@@ -25,6 +26,7 @@ internal class ComponentParametersTypeCache
             }
         }
 
+        [UnconditionalSuppressMessage("ReflectionAnalysis", "IL2026", Justification = "This pattern is not linker friendly.")]
```
Comment (eerhardt) [inline src/Components/Shared/src/ComponentParametersTypeCache.cs:29]: This shouldn't be suppressed. But instead have `RequiresUnreferencedCode` attribute.

File: src/Components/Web/src/Properties/ILLink.Descriptors.xml [hunk 9]
```diff
@@ -0,0 +1,22 @@
+<linker>
+  <assembly fullname="Microsoft.AspNetCore.Components.Web">
+    <type fullname="Microsoft.AspNetCore.Components.Forms.EditForm" required="false" preserve="all" />
```
Comment (eerhardt) [inline src/Components/Web/src/Properties/ILLink.Descriptors.xml:3]: Do we really need the whole Type of all these classes?

File: src/Components/WebAssembly/WebAssembly/src/Prerendering/ClientComponentParameterDeserializer.cs [hunk 10]
```diff
@@ -22,6 +20,8 @@ internal class WebAssemblyComponentParameterDeserializer
 
         public static WebAssemblyComponentParameterDeserializer Instance { get; } = new WebAssemblyComponentParameterDeserializer(new ComponentParametersTypeCache());
 
+        [UnconditionalSuppressMessage("ReflectionAnalysis", "IL2026", Justification = "This pattern is not linker friendly. https://github.com/dotnet/aspnetcore/issues/29946 tracks making parameters linker friendly.")]
```
Comment (eerhardt) [inline src/Components/WebAssembly/WebAssembly/src/Prerendering/ClientComponentParameterDeserializer.cs:23]: `RequiresUnreferencedCode`

File: src/Components/WebAssembly/testassets/WasmLinkerTest/WasmLinkerTest.csproj [hunk 11]
```diff
@@ -0,0 +1,46 @@
...
+    <!--
+        When running from Desktop MSBuild, DOTNET_HOST_PATH is not set.
+        In this case, explicitly specify the path to the dotnet host.
+        -->
+    <PropertyGroup Condition=" '$(DOTNET_HOST_PATH)' == '' ">
+      <!-- This is defined when building in Visual Studio. -->
+      <_DotNetHostDirectory>$(NetCoreRoot)</_DotNetHostDirectory>
+      <_DotNetHostFileName>$([System.IO.Path]::GetFileName('$(DotNetTool)'))</_DotNetHostFileName>
+    </PropertyGroup>
+    <ILLink
```
Comment (eerhardt) [inline src/Components/WebAssembly/testassets/WasmLinkerTest/WasmLinkerTest.csproj:37]: Do you have warnings as errors turned on for this project?

**Comment by eerhardt** [comment]
fyi @vitek-karas @sbomer @MichalStrehovsky

File: src/Components/WebAssembly/testassets/WasmLinkerTest/WasmLinkerTest.csproj [hunk 12]
```diff
@@ -0,0 +1,46 @@
...
+        AssemblyPaths=""
+        RootAssemblyNames="@(RootAssemblies)"
+        OutputDirectory="$(LibrariesTrimmedArtifactsPath)"
+        ReferenceAssemblyPaths="@(RuntimePackAsset);@(ReferencePath->WithMetadataValue('ExternallyResolved', 'true'))"
+        ExtraArgs="$(ILLinkArgs) --verbose"
```
Comment (eerhardt) [inline src/Components/WebAssembly/testassets/WasmLinkerTest/WasmLinkerTest.csproj:42]: You shouldn't need `--verbose`

File: src/Components/Components/src/Reflection/MemberAssignment.cs [hunk 13]
```diff
@@ -11,18 +12,23 @@ namespace Microsoft.AspNetCore.Components.Reflection
     internal class MemberAssignment
     {
         public static IEnumerable<PropertyInfo> GetPropertiesIncludingInherited(
-            Type type, BindingFlags bindingFlags)
+            [DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.PublicProperties | DynamicallyAccessedMemberTypes.NonPublicProperties)] Type type,
+            BindingFlags bindingFlags)
```
Comment (vitek-karas) [inline src/Components/Components/src/Reflection/MemberAssignment.cs:16]: I'm not sure this is enough. The current behavior in linker is that it follow reflection exactly. So this annotation is equivalent to calling type.GetProperties(bindingflags). It does NOT translate to calling the same on the base type. Specifically, private properties on base type are not preserved, but the code here will try to access them.

The reason there's no warning is that linker has a known issue with tracking backward branches, and so the while loop is not correctly analyzed (it sees the `.BaseType` "too late" and doesn't realize that it will call `.GetProperties()` on it).

I think the only "safe" annotation is `All` in this case (which does preserve everything, including private properties on all base types).

Side note: Why do we allow private reflection here... we're telling everybody not to do it, and here we allow it.

File: src/Components/Components/src/Routing/LegacyRouteMatching/LegacyRouteTableFactory.cs [hunk 14]
```diff
@@ -18,6 +19,7 @@ internal static class LegacyRouteTableFactory
             new ConcurrentDictionary<Key, LegacyRouteTable>();
         public static readonly IComparer<LegacyRouteEntry> RoutePrecedence = Comparer<LegacyRouteEntry>.Create(RouteComparison);
 
+        [RequiresUnreferencedCode("This API calls Assembly.ExportedTypes. Types and members of the loaded assembly might be removed during trimming.")]
```
Comment (vitek-karas) [inline src/Components/Components/src/Routing/LegacyRouteMatching/LegacyRouteTableFactory.cs:22]: This will be a public facing warning message. So explaining internal implementation details doesn't help. Ideally it should tell the user what to do about this - call a different api, or something similar.

File: src/Components/WebAssembly/WebAssembly/src/Hosting/WebAssemblyCultureProvider.cs [hunk 15]
```diff
@@ -12,6 +13,7 @@
 
 namespace Microsoft.AspNetCore.Components.WebAssembly.Hosting
 {
+    [UnconditionalSuppressMessage("ReflectionAnalysis", "IL2026", Justification = "This type loads resx files. We don't expect it's dependencies to be trimmed in the ordinary case.")]
```
Comment (vitek-karas) [inline src/Components/WebAssembly/WebAssembly/src/Hosting/WebAssemblyCultureProvider.cs:16]: I admit I don't fully understand what this does, but the justification sounds like "we hope this will work". That's typically not good enough. Maybe we need a different solution here?

File: src/Shared/LinkerFlags.cs [hunk 16]
```diff
@@ -0,0 +1,15 @@
...
+using System.Diagnostics.CodeAnalysis;
+
+namespace Microsoft.AspNetCore.Internal
+{
+    internal static class LinkerFlags
+    {
+        /// <summary>
+        /// Flags for a member that is JSON (de)serialized.
+        /// </summary>
+        public const DynamicallyAccessedMemberTypes JsonSerialized = DynamicallyAccessedMemberTypes.PublicConstructors | DynamicallyAccessedMemberTypes.PublicFields | DynamicallyAccessedMemberTypes.PublicProperties;
```
Comment (vitek-karas) [inline src/Shared/LinkerFlags.cs:13]: These don't apply recursively. So if the type in question is something like:

```C#
class Email { public Attachment Attachment {get; set;} }
class Attachment { public Name {get; set;} }
```

Annotating `Email` with this, will not preserve `Attachment.Name` which the serializer will try to access.
Currently there's no annotation which can be added for serialization purposes. For `System.Text.Json` we basically hope that source generators will solve this.

File: src/Components/Components/src/DefaultComponentActivator.cs [hunk 17]
```diff
@@ -10,10 +11,11 @@ internal class DefaultComponentActivator : IComponentActivator
...
         {
             var instance = Activator.CreateInstance(componentType);
-            if (!(instance is IComponent component))
+            if (instance is not IComponent component)
```
Comment (MichalStrehovsky) [inline src/Components/Components/src/DefaultComponentActivator.cs:18]: This should use Type.IsAssignableFrom to check whether componentType implements the interface **before** activating the type. Otherwise this suppression will be invalid even if illink adds support for the gesture that ensures components are always preserved. If the type is not a component, activator might throw a MissingMethodException because trimming removed the ctor before we reach this check to throw ArgumentException.

The goal of trimming annotations is to never have differences is behaviors at runtime caused by trimming.

File: src/Components/Components/src/Properties/ILLink.Descriptors.xml [hunk 18]
```diff
@@ -0,0 +1,11 @@
+<linker>
```
Comment (MichalStrehovsky) [inline src/Components/Components/src/Properties/ILLink.Descriptors.xml:1]: It would be good to add explanation why these things are unconditionally rooted (e.g. if the type is referenced in any way - a local variable, a never assigned field of this type, etc. it's going to be preserved in its entirety. Usually it's not desirable to have any ILLink.Descriptors.xml and if there are, there should be a good record why.

File: src/Components/Shared/src/ComponentParametersTypeCache.cs [hunk 19]
```diff
@@ -26,7 +26,7 @@ internal class ComponentParametersTypeCache
             }
         }
 
-        [UnconditionalSuppressMessage("ReflectionAnalysis", "IL2026", Justification = "This pattern is not linker friendly.")]
+        [RequiresUnreferencedCode("This type attempts to load component parameters that may be linked away.")]
```
Comment (vitek-karas) [inline src/Components/Shared/src/ComponentParametersTypeCache.cs:29]: The official term is "trim", we try to avoid using "link" in any public facing names/strings since that has typically a different meaning for developers.

File: src/Components/Web/src/Forms/InputTextArea.cs [hunk 20]
```diff
@@ -29,6 +30,7 @@ public class InputTextArea : InputBase<string?>
         [DisallowNull] public ElementReference? Element { get; protected set; }
 
         /// <inheritdoc />
+        [DynamicDependency(BlazorComponent, typeof(InputTextArea))]
```
Comment (vitek-karas) [inline src/Components/Web/src/Forms/InputTextArea.cs:33]: Is it an attempt to workaround a linker solution, or do you want to keep these long-term?

File: src/Components/Components/src/Properties/ILLink.Descriptors.xml (hunk 18 above)
Comment (pranavkm) [reply src/Components/Components/src/Properties/ILLink.Descriptors.xml:1]: I've reduced this list to only include Blazor components. In general, the way these get is not friendly with linker analysis and there isn't an alternate place to put these attributes. https://github.com/mono/linker/issues/1806 would help remove these.

File: src/Components/Web/src/Properties/ILLink.Descriptors.xml (hunk 9 above)
Comment (pranavkm) [reply src/Components/Web/src/Properties/ILLink.Descriptors.xml:3]: These are all components, and preserving these entirely is the safest course of action at this point.

File: src/Components/WebAssembly/WebAssembly/src/Hosting/WebAssemblyCultureProvider.cs (hunk 15 above)
Comment (pranavkm) [reply src/Components/WebAssembly/WebAssembly/src/Hosting/WebAssemblyCultureProvider.cs:16]: Blazor WebAssembly is responsible for it's own assembly loading. With resx files, it waits until after the app has started `Program.MainAsync` has executed to load satellite assemblies based on current culture. These resource assemblies are part of the initial set of assemblies that are trimmed, so we're not loading any assemblies that hadn't participated during linking.

File: src/Components/WebAssembly/testassets/WasmLinkerTest/WasmLinkerTest.csproj (hunk 11 above)
Comment (pranavkm) [reply src/Components/WebAssembly/testassets/WasmLinkerTest/WasmLinkerTest.csproj:45]: Yup, our repo has it on by default.

File: src/Shared/LinkerFlags.cs (hunk 16 above)
Comment (pranavkm) [reply src/Shared/LinkerFlags.cs:13]: That makes sense. These annotations mirror the JsonSerializer's current annotations. Once we have a AOT support there, we'll make sure these APIS are correctly set up to work with the source generator.

File: src/Components/Web/src/Forms/InputTextArea.cs (hunk 20 above)
Comment (pranavkm) [reply src/Components/Web/src/Forms/InputTextArea.cs]: This is a workaround and hopefully something that should be resolved as part of https://github.com/mono/linker/issues/1806. Unfortunately the annotation does not particularly help because of how these APIs get used. I switched back to the linker descriptors for the time being

File: src/Components/Components/src/DefaultComponentActivator.cs [hunk 21]
```diff
@@ -10,15 +11,14 @@ internal class DefaultComponentActivator : IComponentActivator
         public static IComponentActivator Instance { get; } = new DefaultComponentActivator();
 
         /// <inheritdoc />
-        public IComponent CreateInstance(Type componentType)
+        public IComponent CreateInstance([DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.PublicConstructors)] Type componentType)
         {
-            var instance = Activator.CreateInstance(componentType);
-            if (!(instance is IComponent component))
+            if (!typeof(IComponent).IsAssignableFrom(componentType))
```
Comment (SteveSandersonMS) [inline src/Components/Components/src/DefaultComponentActivator.cs:16]: Is the change here in order to improve things for linkability, or just a stylistic preference?

It's probably not a big deal but the implementation before avoided repeated type compatibility checks.

File: src/Components/Components/src/Microsoft.AspNetCore.Components.WarningSuppressions.xml [hunk 22]
```diff
@@ -0,0 +1,47 @@
+﻿<?xml version="1.0" encoding="utf-8"?>
+<linker>
+  <assembly fullname="Microsoft.AspNetCore.Components, Version=6.0.0.0, Culture=neutral, PublicKeyToken=adb9793829ddae60">
+    <attribute fullname="System.Diagnostics.CodeAnalysis.UnconditionalSuppressMessageAttribute">
+      <argument>ILLink</argument>
+      <argument>IL2026</argument>
```
Comment (SteveSandersonMS) [inline src/Components/Components/src/Microsoft.AspNetCore.Components.WarningSuppressions.xml:6]: It's not obvious from reading this file what warnings are being suppressed. What are IL2026 and the others? Could you add comments stating what is the purpose of the suppressions?

File: src/Components/Components/src/Properties/ILLink.Descriptors.xml (hunk 18 above)
Comment (SteveSandersonMS) [reply src/Components/Components/src/Properties/ILLink.Descriptors.xml:1]: Will we need to remember to add all new built-in components in the future to this file? What happens if we don't? Does it just fail to make that new component type eligible for being stripped out?

File: src/Components/Components/src/Reflection/MemberAssignment.cs [hunk 23]
```diff
@@ -11,16 +12,16 @@ namespace Microsoft.AspNetCore.Components.Reflection
...
         {
             var dictionary = new Dictionary<string, List<PropertyInfo>>();
 
             Type? currentType = type;
 
             while (currentType != null)
             {
-                var properties = currentType.GetProperties(bindingFlags)
-                    .Where(prop => prop.DeclaringType == currentType);
+                var properties = currentType.GetProperties(bindingFlags  | BindingFlags.DeclaredOnly);
```
Comment (SteveSandersonMS) [inline src/Components/Components/src/Reflection/MemberAssignment.cs:24]: Nice

File: src/Components/Components/src/Routing/LegacyRouteMatching/LegacyRouteTableFactory.cs [hunk 24]
```diff
@@ -26,7 +27,7 @@ public static LegacyRouteTable Create(IEnumerable<Assembly> assemblies)
                 return resolvedComponents;
             }
 
-            var componentTypes = key.Assemblies.SelectMany(a => a.ExportedTypes.Where(t => typeof(IComponent).IsAssignableFrom(t)));
+            var componentTypes = RouteTableFactory.GetRouteableComponents(key.Assemblies);
```
Comment (SteveSandersonMS) [inline src/Components/Components/src/Routing/LegacyRouteMatching/LegacyRouteTableFactory.cs:30]: Is it possible to leave this file unchanged? The whole idea of the `Legacy*` variants of the routing logic is that they matched an exact point in time that we shipped before, and we keep them decoupled from the newer code so we don't have to reason about how any changes in the new code might impact behaviors in the legacy code.

Eventually (.NET 7?) we'll just remove the whole of the legacy version.

File: src/Components/Components/test/Routing/RouteTableFactoryTests.cs [hunk 25]
```diff
@@ -71,7 +71,7 @@ private class MyComponent : ComponentBase
         public void CanDiscoverRoutes_WithInheritance()
         {
             // Arrange & Act
-            var routes = RouteTableFactory.Create(new[] { typeof(MyComponent), typeof(MyInheritedComponent), });
+            var routes = RouteTableFactory.Create(new List<Type> { typeof(MyComponent), typeof(MyInheritedComponent), });
```
Comment (SteveSandersonMS) [inline src/Components/Components/test/Routing/RouteTableFactoryTests.cs:74]: Curious about why these changes

File: src/Components/Forms/src/Microsoft.AspNetCore.Components.Forms.csproj [hunk 26]
```diff
@@ -12,4 +12,8 @@
     <Reference Include="Microsoft.AspNetCore.Components" />
   </ItemGroup>
 
+  <ItemGroup>
+    <Folder Include="Properties\" />
+  </ItemGroup>
```
Comment (SteveSandersonMS) [inline src/Components/Forms/src/Microsoft.AspNetCore.Components.Forms.csproj:17]: Revert?

File: src/Components/Web/src/Forms/InputFile/InputFileJsCallbacksRelay.cs [hunk 27]
```diff
@@ -13,6 +14,7 @@ internal class InputFileJsCallbacksRelay : IDisposable
 
         public IDisposable DotNetReference { get; }
 
+        [DynamicDependency(nameof(NotifyChange))]
```
Comment (SteveSandersonMS) [inline src/Components/Web/src/Forms/InputFile/InputFileJsCallbacksRelay.cs:17]: This is subtle. Having to reason about all the possible code paths that could result in a given `[JSInvokable]` method to be used. I don't have a better suggestion than what you've done here, but it something we'll have to think hard about in all code reviews that could impact the possible code paths that lead to a JS invokable member being used.

File: src/Components/Web/src/Properties/ILLink.Descriptors.xml (hunk 9 above)
Comment (SteveSandersonMS) [reply src/Components/Web/src/Properties/ILLink.Descriptors.xml:3]: The ones for `M.A.Components` don't have `preserve="all"`. Is the difference just that it's more work (and maintenance cost) to reason about the exact preservability rules for all these extra components in `.Web`?

File: src/Components/Web/src/Virtualization/VirtualizeJsInterop.cs [hunk 28]
```diff
@@ -17,6 +18,8 @@ internal class VirtualizeJsInterop : IAsyncDisposable
 
         private DotNetObjectReference<VirtualizeJsInterop>? _selfReference;
 
+        [DynamicDependency(nameof(OnSpacerAfterVisible))]
+        [DynamicDependency(nameof(OnSpacerAfterVisible))]
```
Comment (SteveSandersonMS) [inline src/Components/Web/src/Virtualization/VirtualizeJsInterop.cs:22]:
```suggestion
        [DynamicDependency(nameof(OnSpacerBeforeVisible))]
```

**Comment by SteveSandersonMS** [comment]
This looks good to me, @pranavkm!

Do you think there's any value in trying to do any specific E2E testing around post-linked applications? I know we have that for the default template currently. Maybe that's enough.

**Comment by javiercn** [comment]
@SteveSandersonMS there's a separate issue for it [here](https://github.com/dotnet/aspnetcore/issues/29903)

File: src/Components/Components/src/DefaultComponentActivator.cs (hunk 21 above)
Comment (pranavkm) [reply src/Components/Components/src/DefaultComponentActivator.cs:16]: I think it might allow the linker to better statically analyze the code particularly if it's unable to piece together the control flow that this is a component type. 

Do you think it's worthwhile removing the type compatibility check (or putting it behind a `Debug.Assert` if it helps the linker analysis) and letting the cast throw? The exception message you get for the invalid cast isn't significantly worse than what you currently get.

File: src/Components/Components/src/Microsoft.AspNetCore.Components.WarningSuppressions.xml (hunk 22 above)
Comment (pranavkm) [reply src/Components/Components/src/Microsoft.AspNetCore.Components.WarningSuppressions.xml:6]: The files were generated by the linker so it'd be a bit of an upkeep to document it. A lot of these are complaining about the various reflection patterns with components and I'm hoping we get a [linker feature](https://github.com/mono/linker/issues/1806) that removes this need.

File: src/Components/Components/src/Properties/ILLink.Descriptors.xml (hunk 18 above)
Comment (pranavkm) [reply src/Components/Components/src/Properties/ILLink.Descriptors.xml:1]: Any unsuppressed linker warning causes a local build to produce a warning when it builds the newly added test project and the CI build to fail because it has WarningsAsErrors=true.

File: src/Components/Components/src/Routing/LegacyRouteMatching/LegacyRouteTableFactory.cs (hunk 24 above)
Comment (pranavkm) [reply src/Components/Components/src/Routing/LegacyRouteMatching/LegacyRouteTableFactory.cs:30]: 👍🏽

File: src/Components/Components/test/Routing/RouteTableFactoryTests.cs (hunk 25 above)
Comment (pranavkm) [reply src/Components/Components/test/Routing/RouteTableFactoryTests.cs:74]: I should undo these. The linker does not flow suppressions to compiler generated state machines. The previous code relied on Linq + yield return which resulted in warnings in the compiler generated code that I did not know how to suppress.

File: src/Components/Web/src/Forms/InputFile/InputFileJsCallbacksRelay.cs (hunk 27 above)
Comment (pranavkm) [reply src/Components/Web/src/Forms/InputFile/InputFileJsCallbacksRelay.cs:17]: You're right. I think `JSInvokable` is one of the few feature that currently has no way to signal a linker warning and you would only find out at runtime. Makes @javiercn's [issue to test these as part of our builds](https://github.com/dotnet/aspnetcore/issues/29903) doubly important.

File: src/Components/Web/src/Properties/ILLink.Descriptors.xml (hunk 9 above)
Comment (pranavkm) [reply src/Components/Web/src/Properties/ILLink.Descriptors.xml:3]: I'll make the two consistent. `preserve="all"` is the default:

```
<!-- No "preserve" attribute and no members specified means preserve all members -->
<type fullname="Assembly.B" /> 
```

https://github.com/mono/linker/blob/main/docs/data-formats.md#preserve-a-type

File: src/Components/Components/src/DefaultComponentActivator.cs (hunk 21 above)
Comment (MichalStrehovsky) [reply src/Components/Components/src/DefaultComponentActivator.cs:16]: It could help analysis, but illink doesn't work that way right now.

The purpose is to ensure that if someone passes a random type here, we don't get a MissingMemberException due to trimming before we would do the type check.

The behavior of the app before/after trimming shouldn't change and that includes error cases like this - they make troubleshooting harder.

Comment (SteveSandersonMS) [reply src/Components/Components/src/DefaultComponentActivator.cs:16]: > Do you think it's worthwhile removing the type compatibility check (or putting it behind a Debug.Assert if it helps the linker analysis) and letting the cast throw? The exception message you get for the invalid cast isn't significantly worse than what you currently get.

I'm certainly fine with letting the error come from the cast if that gives us some linkability benefit. I agree the message on a failed cast exception is clear enough.

File: src/Components/Components/src/Routing/RouteTableFactory.cs [hunk 29]
```diff
@@ -27,13 +28,30 @@ public static RouteTable Create(IEnumerable<Assembly> assemblies)
                 return resolvedComponents;
             }
 
-            var componentTypes = key.Assemblies.SelectMany(a => a.ExportedTypes.Where(t => typeof(IComponent).IsAssignableFrom(t)));
+            var componentTypes = GetRouteableComponents(key.Assemblies);
```
Comment (javiercn) [inline src/Components/Components/src/Routing/RouteTableFactory.cs:31]: Is it Routeable or Routable?

**Review by javiercn — APPROVED** [review]
LGTM

File: src/Shared/LinkerFlags.cs [hunk 30]
```diff
@@ -0,0 +1,20 @@
...
+
+        /// <summary>
+        /// Flags for a component
+        /// </summary>
+        public const DynamicallyAccessedMemberTypes Component = DynamicallyAccessedMemberTypes.All;
```
Comment (marek-safar) [inline src/Shared/LinkerFlags.cs:18]: This looks a bit drastic. Do you really need nested types or private fields?

**Comment by marek-safar** [comment]
Why are Microsoft.JSInterop assemblies not included?

File: src/Shared/LinkerFlags.cs (hunk 30 above)
Comment (pranavkm) [reply src/Shared/LinkerFlags.cs:18]: Blazor relies on being able to reflect on private properties on Components. @vitek-karas suggested using `All` as the most reliable way to preserve these.

**Comment by pranavkm** [comment]
They are imported transitively, but I'll go ahead and add them explicitly
.

**Comment by eerhardt** [comment]
> TBD: Update SDK to trim these assemblies like normal.

@pranavkm - once https://github.com/mono/linker/pull/1839 is flown throughout the system, you can just add `[assembly: AssemblyMetadata("IsTrimmable", "True")]` to these assemblies, and they will be trimmed by the SDK automatically. We can remove all the special case MSBuild targets for these assemblies.

**Comment by pranavkm** [comment]
>  you can just add [assembly: AssemblyMetadata("IsTrimmable", "True")] to these assemblies

Yup, @marek-safar noted this as part of https://github.com/dotnet/aspnetcore/issues/30286
